src/ucp/api/ucp.h

```diff
@@ -481,7 +485,32 @@ enum ucp_am_cb_flags {
      * so the data will be accessible outside the callback, until
      * @ref ucp_am_data_release is called.
      */
-    UCP_AM_FLAG_PERSISTENT_DATA = UCS_BIT(1)
+    UCP_AM_FLAG_PERSISTENT_DATA = UCS_BIT(1),
+
+    /**
+     * Indicates that the data callback may be invoked several times for the
+     * same message, if it was split into several fragments by UCP transfer
+     * protocol. This flag guarantees that the first data callback for a
+     * particular message is invoked for the first fragment. Also the ordering of
+     * first message fragments is guaranteed (i. e. receive callbacks will be called
+     * in the order the messages were sent). The order of other fragments is not
+     * guaranteed. User header is passed with the first fragment only. It is not
+     * allowed to invoke @ref ucp_am_recv_data_nbx function for any of the message
+     * fragments received in this mode of operation. This flag is mutually exclusive
+     * with @a UCP_AM_FLAG_WHOLE_MSG and @a UCP_AM_FLAG_FIRST_MSG flags.
+     */
+    UCP_AM_FLAG_FRAGMENTED_MSG  = UCS_BIT(2),
+
+    /**
+     * Indicates that for multi-fragment eager messages data callback will be invoked
+     * for the first fragment only. Then, to receive the whole message, user has to
+     * invoke @ref ucp_am_recv_data_nbx using pointer to the first fragment data as
+     * data descriptor. The ordering of first message fragments is guaranteed (i. e.
+     * receive callbacks will be called in the order the messages were sent). This
+     * flag is mutually exclusive with @a UCP_AM_FLAG_WHOLE_MSG
+     * and @a UCP_AM_FLAG_FRAGMENTED_MSG flags.
+     */
+    UCP_AM_FLAG_FIRST_MSG       = UCS_BIT(3)
```

Comment:
maybe this could be done automatically if user calls ucp_am_recv_data_nbx for first fragment ?